keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,368 @@
+- open failures with lots of warnings and docs about how old clients might 
+overinterpret the allowed connectivity for a policy
+
+Both of these are impractical due to the nature of security tools, which need to 
+be robust and explicit.  Although failing in a closed manner is acceptable... it is obviously unfortunate,
+as it would obsolete all network policy providers *not* eagerly adopting such a new semantic, which is
+overly restrictive and against the spirit of community-friendly, safe API evolution which is so important
+to a large project like Kubernetes with many external functional plugins.
+
+Ideally, a fix here would “just work” for “back-rev” implementations of NetworkPolicy 
+(which are out-of-tree - generally implemented by cni plugins) and would not 
+introduce new surprises for users. But in practicality this is not the case 
+because, by adding a new selection data structure, the logical inclusion/exclusion 
+properties of namespace and pod selectors become drastically different, as shown in the table below:
+
+
+In this table "policy-reader" might be something like the "calico-controller" or any other cni agent
+that reads the API Servers network policy portfolio and responds to it over time.
+
+|         input policy         | policy-reader (old)    | client (new)        |   |   |
+|------------------------------|------------------------|---------------------|---|---|
+| nsNameSelector + podSelector | ANY ns, pod restricted | ns + pod restricted |   |   |
+| nsNameSelector               | invalid (empty peer)   | ns restricted       |   |   |
+| podSelector                  | ANY namespace          | ANY namespace       |   |   |
+
+In an older client (one which predated the addition of a new, "nsNameSelector" field, we find
+that there is a scenario (`ANY ns, pod restricted`), wherein an old interpretation of the API
+assumes that the ONLY restricting field is a pod label.  Thus, although the user intended to
+restrict all traffic to a namespace/pod combination, the policy implementer does something **much**
+more promiscuous.
+
+This is all avoidable if we could assume, in all cases, that namespaces have a default label.
+
+There may also be many other usecases where the a security or permissions boundary might be conveniently defined in terms
+of a namepsace name, and that such a definition would be most easily referenced by a default label.
+
+### Goals
+
+- Add an immutable default label to ALL namespaces, as the namespace name so this 
+can be used as selector by arbitrary components.
+
+### Non-Goals
+
+* Publishing name labels on all resources
+* Exposing arbitrary fields as selectors
+
+## Proposal
+
+We propose to label of all namespaces, by default, with a reserved label 
+(i.e. “kubernetes.io/metadata.name” or some such), to allow easy selection.
+
+If this label is missing, it is added by the apiserver
+
+If the label is incorrect or changed, the apiserver fails to validate the object
```

Comment:
this is probably a remote and maybe even unlikely possibility, what happens to namespaces, prior to upgrade, who had same label key(whatever we choose) but a different value.. these namespaces will now be invalidated which previously worked fine.

Reply:
Good question -- we own `kubernetes.io/` prefixed labels so I don't think we are violating any compatibility rules if a user was using it for some other use-case outside of what Kubernetes is managing.

Reply:
thanks Andrew. I am aware of this. However we don't validate against it so it remains a possibility. Only reason i raised this point is to clarify the approach that will be taken for such a case.

Reply:
Good point -- we should at least write down what the expected behavior for that user would be. It sounds like worse case scenarios is we disallow any future writes to that namespace object.

Reply:
This is why we should have one alpha release where we warn if we see people using it

Reply:
seems like we should , in general make this  label EXTREMELY specific so that this just **doesnt** happen 
  `k8s.io/apimachinery.namespace.default.name` (alternatively , literally plumb it in as an apiserver option that can be overriden if folks want to)

Reply:
i think the name proposed in the KEP is perfectly fine IMO..

Reply:
Ok happy to keep the name as is , resolving

Reply:
updated text

Reply:
First question is whether we want to fail if someone tries to set the value.  That was my initial feeling, but it sounds like you are going full-tilt.  I'm OK with that, I guess.  It does answer "what about pre-existing".

The simple answer seems to be to just blindly set the value.  I think defaults is good enough, isn't it?  The only downsides is that it is versioned.  Doing it in PrepareFor and Decorator is more involved.  Can I select on labels that only are set in Decorator?

Reply:
> First question is whether we want to fail if someone tries to set the value. That was my initial feeling, but it sounds like you are going full-tilt. I'm OK with that, I guess. It does answer "what about pre-existing".

I like my invariants invariant

> The simple answer seems to be to just blindly set the value. I think defaults is good enough, isn't it?

Though it's unorthodox for defaults to stomp existing values, that would certainly be the simplest place to do it (done in a single place, applies to reading existing data and accepting create/update requests). We could go belt-and-suspenders with a validation check that requires the label exist and match the name (so if we forget to add defaulting for "v2" namespaces it would be caught).

The only thing stomping in defaults would not let us do would be to detect/warn if a user tried to set the label to some other value (since by the time we got to the strategy the conflicting user-specified value would be gone). That seems unlikely to be done accidentally, so I'm not sure how much I care about that.

Reply:
from what i gather, both of you sound _OKAY_ with "default stomping" except that it does not allow us a chance to warn users, but coupled with the fact that the label actually belongs to a reserved prefix of "kubernetes.io/", its okay to override such (ill-advised) usage of these labels?

Reply:
Ha, I had two thoughts here:

1. After this feature is enabled, will label selectors work for Namespaces, or will there need to be a read-write cycle on each Namespace object (like with CRDs for version upgrades)?

2. Would it make sense to use/reserve `metadata.kubernetes.io/` as a prefix for this kind of usage (reflecting field values to labels) without any promises of future reflections?

Reply:
> 1\. After this feature is enabled, will label selectors work for Namespaces, or will there need to be a read-write cycle on each Namespace object (like with CRDs for version upgrades)?

this would work on read, no write cycle required

> 2\. Would it make sense to use/reserve `metadata.kubernetes.io/` as a prefix for this kind of usage (reflecting field values to labels) without any promises of future reflections?

I'd tend to keep the `metadata.name` construct together rather than splitting it

Reply:
I LOVE idea # 2 because @deads2k gently suggested that, why not reflect other scalar api fields

Reply:
this thread is long: oversimplifying the comments:  


```
@thockin: why we should have an alpha release
@liggit: reasonable to make use of kube.io labels w/o an intermediate release
         strongly force the label value
          - add if absent
          - prepareForCreate/PrepareForUpdate ~ override wrong values
          - Decorate: replace differeing values
@thockin is ok with ^, but "Can I select on labels that only are set in Decorator?"
@evan reserve `metadata label` + also wants "add if absent"

```

Reply:
Re name: <something.>kubernetes.io/metadata.name seems best - extends well and maps to field-selector syntax.  The sentiment on "<something.>" seems to be "", which I am fine with.

Re alpha: I am not convinced - I still think a cycle of warning is "the polite thing to do".  After that, I am OK with unilateral stomping.

Re where: I tested defaults and it seems to work.  I tested Decorator and it does NOT seem to work.  This actually seems like a problem to me, but not one for today.